cpp/harvest/marc_issn_lookup.cc

```diff
@@ -99,86 +152,123 @@ void UpdateSubfield(MARC::Subfields &subfields, const SubFieldInfo &sub_field_in
     if (!subfields.replaceFirstSubfield('w', sub_field_info.w_))
         subfields.addSubfield('w', sub_field_info.w_);
 
-    if (not sub_field_info.t_.empty()) {
+    if (not sub_field_info.t_.empty())
         if (!subfields.replaceFirstSubfield('t', sub_field_info.t_))
             subfields.addSubfield('t', sub_field_info.t_);
+}
+
+
+void UpdateSubFieldInfo(SubFieldInfo &sfi, const SubFieldInfo &new_sfi, const bool is_online) {
+    sfi.t_ = new_sfi.t_;
+    sfi.w_ = new_sfi.w_;
+    sfi.x_ = new_sfi.x_;
+    sfi.is_online_ = is_online;
+    InsertIfNotExist(new_sfi.issns_, &sfi.issns_);
+}
+
+
+void UpdateSubFieldAndCombineIssn(SubFieldInfo * const sfi, const SubFieldInfo &new_sfi) {
+    bool subset(true);
+
+    for (const auto &v1 : new_sfi.issns_)
+        subset = (subset || IsInISSNs(sfi->issns_, v1));
+
+    if (not subset) {
+        if (sfi->is_online_ && new_sfi.is_online_) {
+            sfi->is_valid_ = false;
+        }
+
+        if (!sfi->is_online_ && new_sfi.is_online_) {
+            sfi->is_valid_ = true;
+            sfi->t_ = new_sfi.t_;
+            sfi->w_ = new_sfi.w_;
+            sfi->x_ = new_sfi.x_;
+        }
+
+        if (!sfi->is_online_ && !new_sfi.is_online_)
+            sfi->is_valid_ = false;
+
+        InsertIfNotExist(new_sfi.issns_, &sfi->issns_);
     }
 }
 
 
-void UpdateJournalValidity(std::map<std::string, SubFieldInfo> &journal_cache) {
-    // check whether the issn is occur more than once
-    unsigned online_valid(0), online_invalid(0), print_valid(0), print_invalid(0);
-
-    LOG_INFO("Cache contents:");
-    for (auto sfi : journal_cache) {
-        if (sfi.second.online_version_counter_ == 1) {
-            journal_cache[sfi.first].is_valid_ = true;
-            LOG_INFO(sfi.first + ": Online (valid)");
-            ++online_valid;
-        } else if (sfi.second.online_version_counter_ > 1) {
-            journal_cache[sfi.first].is_valid_ = false;
-            LOG_INFO(sfi.first + ": Online (invalid)");
-            ++online_invalid;
-        } else if (sfi.second.online_version_counter_ == 0)
-            if (sfi.second.printed_version_counter_ == 1) {
-                journal_cache[sfi.first].is_valid_ = true;
-                LOG_INFO(sfi.first + ": Print (valid)");
-                ++print_valid;
-            } else {
-                journal_cache[sfi.first].is_valid_ = false;
-                LOG_INFO(sfi.first + ": Print (invalid)");
-                ++print_invalid;
+std::vector<SubFieldInfo> MergeIssn(const std::vector<SubFieldInfo> &journal_cache) {
```

Comment:
1) Instead of passing a reference and creating a copy, could we instead just pass a pointer and do the operations directly on the cache? (it can have multiple thousand entries and might be slow to work on a copy)

2) MergeIssn => maybe we could find a better function name like e.g. `MergeDuplicateSubFieldInfos`